File: rainbowtable/reduction.py
import random
import string
import sys


def reduction_function1(hash_str: str, length: int, index: int) -> str:
    """
    A function to reduce a given hash to a plaintext password of the given length.

    Based on https://link.springer.com/chapter/10.1007/978-3-642-30436-1_42
    """
    if length < 1:
        raise ValueError('length must be greater 0')
    if index < 0:
        raise ValueError('index must be greater or equal 0')

    # Convert hash from string into int
    number = int(''.join(map(str, map(ord, hash))))
    number = (number + index) % 26**length
    result = ''
    for _ in range(0, length):
        result += chr((number % 26) + ord('a'))
        number = number // 26

    return result


def reduction_function2(hash_str: str, length: int, index: int) -> str:
    """
    A function to reduce a given hash to a plaintext password of the given length.

    Based on https://link.springer.com/chapter/10.1007/978-3-642-30436-1_42
    """
    if length < 1:
        raise ValueError('length must be greater 0')
    if index < 0:
        raise ValueError('index must be greater or equal 0')

    if len(hash_str) % 2 == 1:
        hash_str += random.choice(string.digits + 'abcdef')

    # Convert hash from string into int
    number = int.from_bytes(bytes.fromhex(hash_str), byteorder='big', signed=False)
    # number = (number + index) % 26**length
    number = number % 26**length
    result = ''
    for _ in range(0, length):
        result += chr((number % 26) + ord('a'))
        number = number // 26

    return result


def reduction_function3(hash_str: str, length: int, index: int) -> str:
    if len(hash_str) % 2 == 1:
        hash_str += random.choice(string.digits + 'abcdef')

    hash_bytes = bytes.fromhex(hash_str)
    i = int.from_bytes(hash_bytes, byteorder=sys.byteorder)

    result_string = ""
    alphabet = string.ascii_lowercase
    base = len(alphabet)
    while i:
        i, idx = divmod(i, base)
        result_string = alphabet[idx:idx+1] + result_string
    return result_string[:length]

File: rust/rainbowtable/src/rainbow_table/reduction.rs
use std::ops::Add;

use custom_error::custom_error;
use num::bigint::ParseBigIntError;
use num::{BigUint, FromPrimitive, Integer, ToPrimitive};

custom_error! {pub ReductionError
  Format{source: ParseBigIntError} = "some input is of invalid format",
  Other{message: String} = "{message}"
}

/// A function to reduce a given hash to a plaintext password of the given length.
///
/// Based on https://link.springer.com/chapter/10.1007/978-3-642-30436-1_42
pub fn reduction_function1(
    hash_str: &str,
    length: u32,
    index: u32,
) -> Result<String, ReductionError> {
    // number = int(''.join(map(str, map(ord, hash_str))))

    // Generate an iterator where each value is the unicode point of each char in the hash string.
    let step_n_3 = hash_str.chars().map(|c| c as u8);
    // Convert each of the unicode points into a string.
    let step_n_2 = step_n_3.map(|byte| byte.to_string());
    // Join them together to a single string.
    let step_n_1 = step_n_2.collect::<Vec<String>>().join("");
    // Parse the string as the big int it might be.
    let step_n = step_n_1.parse::<BigUint>()?;

    // number = (number + index) % 26 ** length
    let num_26 = match BigUint::from_i32(26) {
        Some(n) => n,
        // This shouldn't possibly happen.
        None => {
            return Err(ReductionError::Other {
                message: "26 into bigint failed".into(),
            })
        }
    };
    let mut number = step_n + index % num_26.pow(length);

    // result = ''
    // for _ in range(0, length):
    //     result += chr((number % 26) + ord('a'))
    //     number = number // 26
    let mut result = String::new();
    for _ in 0..length {
        let (quotient, remainder) = number.div_rem(&num_26);
        // ord('a') == 97
        let codepoint = match (remainder + 97u32).to_u32() {
            Some(u) => u,
            None => {
                return Err(ReductionError::Other {
                    message: "codepoint error".into(),
                })
            }
        };
        let char = match char::from_u32(codepoint) {
            Some(char) => char,
            None => {
                return Err(ReductionError::Other {
                    message: "char from codepoint failed".into(),
                })
            }
        };
        result = result.add(&char.to_string());
        number = quotient;
    }

    return Ok(result);
}
